x-pack/plugins/files/server/file_share_service/generate_share_token.ts

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import crypto from 'crypto';
+
+/**
+ * This code is adapted and borrowed from
+ * https://github.com/sindresorhus/crypto-random-string/blob/25f8930282bc3803d03b9080e710db8426430b63/core.js
+ */
+
+const CHAR_SET = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
+const CHAR_SET_LENGTH = CHAR_SET.length;
+
+// We are going to read as 16 bits numbers at a time which means this is the largest
+// number we are going to see.
+const MAX_16_BIT_INT = Math.pow(2, 16) - 1; // OR 0x10000 - 1
+
+// The biggest number we can use for modulo given the charset to ensure good distribution
+// This number works because we know CHAR_SET_LENGTH fits into it CHAR_SET_LENGTH number of times
+const MAX_USABLE_VALUE = Math.floor(MAX_16_BIT_INT / CHAR_SET_LENGTH) * CHAR_SET_LENGTH;
+
+// Tokens must be 40 chars long
+const TOKEN_LENGTH = 40;
+// How much entropy we will generate for each pass
+const ENTROPY_LENGTH = 2 * Math.floor(1.1 * TOKEN_LENGTH);
+
+/**
+ * Generate 40 random characters of a pre-checked charset for share tokens.
+ *
+ * Samples:
+ *
+ * XBWUuHRZsxXZgS03VUp2hSifmCoF4zznsNOehTea
+ * 40RIzynCs2OFTkYC2fy9vfLgYhxjmwzYrJvGj2Ih
+ * ZFM1vFSz6k8k9QU4Ve5mu0PbcBi3rwV4BykbcrDU
+ * 2oweE4dmWT7warVYA59l0thCDWnIMvUsAC9yCWyQ
+ * q8R4LneRCBhcsCNKgcebZRh1Mh4m746GNTuifmAA
+ * kpUzKyynu8uXDtek2Xg5yJRwPkw37Cd1VaUezNZ2
+ */
+export function generateShareToken(): string {
+  let token = '';
+  while (token.length < TOKEN_LENGTH) {
+    // It is highly unlikely we will call this function more than once per token
+    const entropy = crypto.randomBytes(ENTROPY_LENGTH);
+    let entropyPointer = 0;
+    while (entropyPointer < entropy.length && token.length < TOKEN_LENGTH) {
+      const value = entropy.readUInt16LE(entropyPointer);
+      entropyPointer += 2; // Move ahead 2 bytes to the next number
+      if (value > MAX_USABLE_VALUE) continue;
+      token += CHAR_SET[value % CHAR_SET_LENGTH];
```

Comment:
Given that `CHAR_SET_LENGTH` is less than 255, I guess we can use one octet at a time from the `entropy` buffer.

```suggestion
      const value = entropy.readUInt8(entropyPointer);
      entropyPointer += 1;
      if (value > MAX_USABLE_VALUE) continue;
      token += CHAR_SET[value % CHAR_SET_LENGTH];
```

Reply:
I'm not sure I follow? Reading 16-bit numbers gives us larger numbers which improves the randomness of each char we choose. I'm not sure how it relates to CHAR_SET_LENGTH?

Reply:
It relates to `CHAR_SET_LENGTH`, because that is the length of possible chars we use, lets say it is 64 (64 possible characters). So, we need 6-bits of entropy to pick a character. One octet (byte) already gives us 8-bits of entropy, so more than we need. Thus my comment (question) is: why do we pick 2 octets&mdash;16 bits of entropy&mdash;when we need only 6 bits?

Reply:
I guess the algorithm in `crypto-random-string` does it to support encoding in different bases, including in bases larger than 255. But here we don't need to support different char sets, we can just pick, say Base64 or Base36, or Base16.

Reply:
Ah ok I see what you're saying. You mean map directly from random bits to a char in the charset. This algo is not doing that, not directly at least, because of how it maps randomness to chars using modulo.

But I think you have a good point about simplifying it to our specific use case using Base36! Let's use that instead.